walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -13,11 +13,157 @@
 // limitations under the License.
 
 package com.google.sps;
-
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.ArrayList;
 import java.util.Collection;
 
+/**
+ * Class which finds a range of times when all required attendees can meet
+ * If there are no available times when an all optional attendees can meet,
+ * then only required attendees are considered in finding available times 
+ */
 public final class FindMeetingQuery {
+
+  /**
+   * Find available time ranges to meet
+   * @param events All events which are occuring 
+   * @param request A requested meeting with certain required and optional attendees and a duration
+   * @return all time ranges when the required attendess, and possibly the optional attendees, can meet
+   */
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    Collection<String> allAttendees = new ArrayList<String>();
+    allAttendees.addAll(request.getAttendees());
+    allAttendees.addAll(request.getOptionalAttendees());
+    Collection<TimeRange> openTimes = queryWithAttendees(events, request, allAttendees);
+    if (request.getAttendees().size() == 0 || openTimes.size() != 0)
+      return openTimes;
+    return queryWithAttendees(events, request, request.getAttendees());
```

Comment:
nit: I think it would be helpful to leave a comment above this return statement to clarify that this will return open times for non-optional attendees to help differentiate it from the queryWithAttendees call above